### PR TITLE
Fixed __hash__ of range

### LINKED ISSUE #1482: hash of range object is not deterministic.
## Feature
Currently, hash of range is not deterministic. 
The range's `__hash__` is not implemented, so the object's `__hash__` is called. (it returned id of object)

### Expected result
```python
>>> hash(range(10))
-6299899980521991026
>>> hash(range(10))
-6299899980521991026
>>> a = {}
>>> a[range(10)] = 1
>>> a[range(10)] = 1
>>> a
{range(0, 10): 1}
```

### Actual result
```python
>>>>> hash(range(10))
140482322810896
>>>>> hash(range(10))
140482319976752
>>>>> a = {}
>>>>> a[range(10)] = 1
>>>>> a[range(10)] = 1
>>>>> a
{range(0, 10): 1, range(0, 10): 1}
```

### PR DESCRIPTION
**Pull request by HyeockJinKim**
Calculate hash value of range through hash of tuple

Fixes #1482

### CODE REVIEW

File: vm/src/obj/objrange.rs [hunk 1]
```diff
@@ -402,6 +404,25 @@ impl PyRange {
...
+        let mut c = vm.get_none();
+        if !len.is_zero() {
+            b = zelf.start.clone().into_object();
+            if !len.is_one() {
+                c = zelf.step(vm).into_object();
+            }
+        };
+        let tuple = PyTuple {
+            elements: vec![a, b, c],
+        };
```
Comment (HyeockJinKim) [inline vm/src/obj/objrange.rs]: I'm not sure if PyTuple creation is the way above.
Is there a better way?

**Comment by HyeockJinKim** [comment]
@coolreader18 @youknowone cc @corona10 
Can I get a review?

File: vm/src/obj/objrange.rs (hunk 1 above)
Comment (youknowone) [reply vm/src/obj/objrange.rs]: looking for `vm.ctx.new_tuple`?

**Comment by youknowone** [comment]
note for reviewer: cpython range hash https://github.com/python/cpython/blob/master/Objects/rangeobject.c#L462-L510

File: vm/src/obj/objrange.rs [hunk 2]
```diff
@@ -402,6 +404,27 @@ impl PyRange {
         }
     }
 
+    #[pymethod(name = "__hash__")]
+    fn hash(zelf: PyRef<Self>, vm: &VirtualMachine) -> PyResult<pyhash::PyHash> {
+        let length = zelf.length();
```
[the remark this answers is not in the document]
Comment (HyeockJinKim) [reply vm/src/obj/objrange.rs]: The code above looks better, so I modified it in the above way.

File: vm/src/obj/objrange.rs (hunk 1 above)
Comment (coolreader18) [reply vm/src/obj/objrange.rs]: It looks like PyTuple's hash just uses `pyhash::hash_iter`, so you could probably do the same.

File: vm/src/obj/objtuple.rs [hunk 3]
```diff
@@ -179,7 +179,7 @@ impl PyTuple {
     }
 
     #[pymethod(name = "__hash__")]
-    fn hash(&self, vm: &VirtualMachine) -> PyResult<pyhash::PyHash> {
+    pub fn hash(&self, vm: &VirtualMachine) -> PyResult<pyhash::PyHash> {
```
Comment (coolreader18) [inline vm/src/obj/objtuple.rs]: Oh, and now this doesn't have to be `pub`.

**Comment by windelbouwman** [comment]
Nice!

**Pull request merged**